jorvik/celery.py

```diff
@@ -0,0 +1,64 @@
+from __future__ import absolute_import
+
+import os
+import traceback
+from celery import Celery, group
+from celery.utils.log import get_task_logger
+from django.db import transaction
+
+os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'jorvik.settings')
+
+app = Celery('jorvik')
+app.config_from_object('django.conf:settings', namespace='CELERY')
+
+# Load task modules from all registered Django app configs.
+app.autodiscover_tasks()
+
+logger = get_task_logger(__name__)
+
+
+# noinspection PyUnusedLocal
+@app.on_after_configure.connect
+def setup_periodic_tasks(sender, **kwargs):
+    sender.add_periodic_task(10.0, process_queue.s(500), name='email queue processor')
+
+
+@app.task(ignore_result=True)
+def send(pk):
+    # noinspection PyBroadException
+    try:
+        from posta.models import Messaggio
+        return Messaggio.invia(pk)
+    except Exception as e:
+        logger.error(traceback.format_exc())
+        return ['FAIL: worker {}'.format(e)]
+
+
+def get_email_batch(limit=None):
+    from posta.models import Messaggio
+    with transaction.atomic():
+        # purtroppo skip_locked è stata inserita solo dalla version 1.11
+        query = '''SELECT id FROM posta_messaggio WHERE terminato is NULL ORDER BY priorita'''
```

Comment:
Non sono molto convinto dall'idea di utilizzare SQL, specialmente nel dialetto di PostgreSQL. Siamo riusciti a non scrivere neanche una riga di SQL per decine di migliaia di righe di codice, preferirei non iniziassimo ora :) Inoltre, sono preoccupato dalle interazioni di Django con il database mentre il record e' bloccato a livello di database (gli amministratori non potrebbero facilmente interagire col messaggio mentre questo e' bloccato, dal pannello di Django). E cosa succede al LOCK in caso di riavvio del sistema?

Mi piace molto l'idea del task Celery per lo smistamento della posta, mi chiedo se non sia il caso di portare fino in fondo questa idea ed affidare il compito di bloccare/sbloccare il messaggio a Celery stesso, piuttosto che farlo noi col database.

Per esempio, potremmo avviare il task del messaggio (`send(pk)`) direttamente dal metodo `costruisci_e_accoda`. Nell'avviare un task, Celery ci restituirebbe l'ID del task, che potremmo salvare atomicamente in database per quel messaggio (es. `Messaggio.task_id`). In questo modo, avremmo un task Celery per ogni messaggio - il blocco e lo sblocco verrebbe gestito direttamente da Celery.

Allo stesso tempo, ci sarebbe comunque il rischio che un task celery non venga eseguito o si interrompa bruscamente per qualche errore. In quel caso, un task periodico (come `process_queue`) potrebbe cercare tutti i messaggi che non sono ancora terminati (es. `terminato__isnull=True`). E' possibile usare l'ID del task (es. `Messaggio.task_id`) per chiedere a Celery qual e' lo stato del task in coda (http://docs.celeryproject.org/en/latest/reference/celery.result.html, vedi `AsyncStatus.result`). Se il task per quel messaggio e' fallito, oppure assente, si puo' riavviare il task (dovremmo pero' tenere traccia del numero di tentativi rimanenti, es. `Messaggio.tentativi_rimanenti`, in modo da non provare all'infinito).

Reply:
I lock terminano all'uscita del blocco `with transaction.atomic()` questo comprende anche il kill del processo. Se da admin provi a modificare il record, il browser ti rimane in attesa fino a quando il record non viene sbloccato, quando il lock sparisce l'admin salva il record perdendo le modifiche della `send`, questo non è evitabile in ogni caso.

Django 1.11 integra `skip_locked()`, non potendo aggiornare purtroppo ho dovuto fare rawquery, e sì non è portabile (psql 9.5+ only).

Per fare due funzioni di invio bisogna inserire un altro campo `in_coda` per evitare che lo spooler tenti di inviare il messaggio, teoricamente se la creazione e l'invio avvenisse nella stessa transazione non servirebbe, però non so se è semplice strutturare così la cosa.

Utilizzare celery con i result invece di psql con skip locked secondo me è meno solida come soluzione, inoltre aggiunge dell'overhead (abbiamo una tabella aggiuntiva con i risultati), altra problematica è che celery non permette di avere il task periodico unico quindi se non vogliamo usare skip_locked() di postgres 9.5 è obbligatorio impedire esecuzioni multiple della funzione `process_queue` utilizzando espedienti simili (questo usa memcached, un altro pezzo ancora):

http://docs.celeryproject.org/en/latest/tutorials/task-cookbook.html#cookbook-task-serial

oppure un lock su una tabella apposita.

Reply:
> Utilizzare celery con i result invece di psql con skip locked secondo me è meno solida come soluzione, inoltre aggiunge dell'overhead (abbiamo una tabella aggiuntiva con i risultati),

Mi sono espresso male, non suggerivo una nuova tabella, solo un campo nella tabella `Messaggio` per il `task_id`.

> altra problematica è che celery non permette di avere il task periodico unico quindi se non vogliamo usare skip_locked() di postgres 9.5 è obbligatorio impedire esecuzioni multiple della funzione process_queue utilizzando espedienti simili (questo usa memcached, un altro pezzo ancora):

Non credo sia un problema impedire esecuzioni multiple. Se appena creato il task aggiorniamo immediatamente il campo `task_id` del relativo `Messaggio`, quando una seconda esecuzione passera' dal `Messaggio` lo trovera' in uno dei due stati: senza `task_id`, quindi da rischedulare, oppure con un `task_id`, quindi da ignorare. 

Se abbiamo paura che nell'arco di tempo tra quando abbiamo effettuato la query, a quando si stiamo iterando sul singolo elemento, sia cambiato lo stato del messaggio, non credo sia necessaria una intera soluzione di locking, ma possiamo magari usare qualcosa come [`Model.refresh_from_db`](https://docs.djangoproject.com/en/2.0/ref/models/instances/#refreshing-objects-from-database), es.

```python
for msg in Messaggio.objects.filter(terminato__isnull=True,
                                    tentativi_rimanenti__gt=0):
    msg.refresh_from_db()
    task_id = msg.task_id
    # controlla status di task con id task_id
    if status in (PENDING, STARTED, SUCCESS):
        continue  # ignora
    msg.invia()
    # ...

class Messaggio:
    def invia(self):
        self.task_id = crea_task()  # atomicamente
        self.save()
```

> Django 1.11 integra skip_locked(), non potendo aggiornare purtroppo ho dovuto fare rawquery, e sì non è portabile (psql 9.5+ only).

Capisco, concordo che `skip_locked()` sarebbe ideale. Non sono sicuro pero' che valga la pena usare SQL.

> Per fare due funzioni di invio bisogna inserire un altro campo in_coda per evitare che lo spooler tenti di inviare il messaggio, teoricamente se la creazione e l'invio avvenisse nella stessa transazione non servirebbe, però non so se è semplice strutturare così la cosa.

OK. Dobbiamo prevedere il caso dell'invio immediato. Le API di `Messaggio` promettono gia' questa funzionalita', vari componenti vi fanno affidamento (es. il modulo di autenticazione), non possiamo rimuoverla.

Reply:
L'invio immediato è stato implementato (così come le altre richieste), per quanto riguarda la gestione coda vorrei invitarvi a leggere questo post:
https://blog.2ndquadrant.com/what-is-select-skip-locked-for-in-postgresql-9-5/

Reply:
> L'invio immediato è stato implementato (così come le altre richieste),

Grazie mille.

> per quanto riguarda la gestione coda vorrei invitarvi a leggere questo post:
https://blog.2ndquadrant.com/what-is-select-skip-locked-for-in-postgresql-9-5/

Ho letto, mi piace molto, ma continuo a non credere che sia abbastanza da legittimare il fare una eccezione sulle linee guida per lo sviluppo,

> A.3. Il database utilizzato è PostgreSQL versione 9.5. Tutto il codice deve essere compatibile con le versioni 9.4 e 9.5. Ragionevoli sforzi dovrebbero essere compiuti per mantenere il software indipendente da ogni particolare dialetto SQL

Dal 2012 siamo gia' passati da 4 DBMS diversi (MySQL, MariaDB, MongoDB, PostgreSQL). E' stata proprio la (terribile :face_with_head_bandage:) esperienza con la portabilita' tra questi database (e le migrazioni di dati) che ci hanno portato ad inserire quel punto nelle linee guida.

Reply:
@acarmisciano ecco una serie di altre alternative che possono fornire la garanzia che cerchiamo:
- Usare una coda Celery separata per il task periodico. Visto che l'unico compito del task periodico e' quello di mettere le email in coda, e' accettabile che questo task venga processato da un singolo worker. Il singolo worker puo' essere configurato per lavorare su un singolo task per volta, evitando quindi problemi di esecuzione multipla contemporanea.
- Modificare il task periodico in modo che crei un file di lock quando viene avviato, e lo distrugga quando viene terminato e va in errore (utilizzando lo storage condiviso in rete). Il task periodico controlla che il file non esista gia' prima di inziare a lavorare. Se vogliamo evitare che il file rimanga in seguito a un errore imprevedibile (kernel panic, corrente, ecc.) possiamo usare uno dei vari metodi classici:
  - Scrivere il PID nel file di lock, e verificare che il processo sia ancora attivo (questo assume che i task periodici vengono eseguiti su un solo worker),
  - Scrivere un timestamp nel file di lock, e ignorarlo se e' passato un certo tempo (configurabile)